### DISCUSSION TITLE
How to compute second derivative in q-points

### DISCUSSION THREAD

**Discussion by andrey1887** [idaholab/moose#26146] [Q&A Modules: Solid mechanics]
Help me please. I need to compute second derivative in q-points. It's line element with Hermite third order shape function.

**Comment by GiudGiud** [comment]
Hello

In the coupleable interface you should find a routine to get the Hessian. Maybe it s called coupledHessian I can check tomorrow evening 

Guillaume

**Reply by GiudGiud** [reply]
Actually it s called coupledSecond

https://mooseframework.org/getting_started/new_users.html#create-an-app

**Reply by andrey1887** [reply]
Thank you, @GiudGiud. I would like to clarify: is it suitable for the next case. In `ComputeIncrementalBeamStrain::computeQpStrain()` a gradients of displacements and rotations are computed:
```
const RealVectorValue grad_disp_0(1.0 / _original_length[0] * (_disp1 - _disp0));
const RealVectorValue grad_rot_0(1.0 / _original_length[0] * (_rot1 - _rot0));
const RealVectorValue avg_rot(
      0.5 * (_rot0(0) + _rot1(0)), 0.5 * (_rot0(1) + _rot1(1)), 0.5 * (_rot0(2) + _rot1(2)));
```
I consider it like as the first derivatives. If we have 1 q-point it is appropriate solution.
But to implement beam model based on Euler-Bernoulli theory, where shear deformations are neglected, the second derivative of rotation is needed to calculate shear forces Q =  EIθ".
So i need to use Hermite shape functions. And then, in `computeQpStrain()` instead of existing gradient calculation with linear shape function for transverse displacements and rotations:
<img src="https://github.com/idaholab/moose/assets/114737834/02865ade-d60f-4b01-9064-0ba6a0bc0e6a">


I will to define third order Hermite shape function manually, to retrieve q-point values, and also first and second derivatives 
<img src="https://github.com/idaholab/moose/assets/114737834/77aca07e-89fb-45bd-95bb-e87fde8c5263" width="400">

Is there a more appropriate way to obtain second derivatives for this case, than manual processing of shape function.

**Reply by GiudGiud** [reply]
I would try `coupledSecond` first. This is the cannon way of retrieving second derivatives. And if there's issues with either Hermite shape functions or beam elements we should know and fix them

any reason you cannot just increase the order on the shape functions you were using previously instead of switching to Hermite?

**Reply by andrey1887** [reply]
Thank you @GiudGiud. A Hermite's shape functions are needed because they provide С1 continuity. 
Could gradients in the snippet below be calculated using `coupledGradient`?
https://github.com/idaholab/moose/blob/d62c3f007814fc404a41ae422d00408da9a24ba7/modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C#L241-L245

**Reply by GiudGiud** [reply]
ah probably not otherwise they would have been.
and that s likely because _disp1 is not a MOOSE variable, just a class attribute that gets set. 

Then you are right, you wont be able to use coupledSecond. You ll want to compute the second derivative manually just like they computed the gradients manually

**Reply by andrey1887** [reply]
Ok. Thank you @GiudGiud.

**Comment by andrey1887** [comment]
So, a Hermit third order shape functions are needed, but if i try to change even one of variables:

```
  [disp_x]
    order = FIRST
    family = HERMITE
  []
```

 i get the next error: 
```
*** ERROR ***                                                                                                                         
std::bad_alloc                                                                                                                        
 
Stack frames: 7
0: libMesh::print_trace(std::ostream&)
1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
2: void mooseError<char const*>(char const*&&)
3: /home/and/projects/moose/framework/libmoose-opt.so.0(+0x92678b) [0x7fab9fdeb78b]
4: main
5: __libc_start_main
6: ../thin_walled-opt(+0x33ef) [0x55a3d35673ef]
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
```
Tell me please: What is the right way to use Hermite shape functions for beam?

**Reply by GiudGiud** [reply]
Only 3rd order hermite are supported iirc

**Reply by andrey1887** [reply]
@GiudGiud , thank you. But now i have another error:
```
30 Nonlinear |R| = 2.205375e+01
Nonlinear solve did not converge due to DIVERGED_MAX_IT iterations 30
 Solve Did NOT Converge!
Aborting as solve did not converge


*** ERROR ***                                                                                                                         
The following error occurred in the object "ConstantDT", of type "ConstantDT".                                                        
                                                                                                                                      
Solve failed and timestep already at or below dtmin, cannot continue!                                                                 

Stack frames: 12
0: libMesh::print_trace(std::ostream&)
1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
2: callMooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, MooseApp*)
3: /home/and/projects/moose/framework/libmoose-opt.so.0(+0x9c4301) [0x7fa7596f9301]
4: /home/and/projects/moose/framework/libmoose-opt.so.0(+0x1bf45d7) [0x7fa75a9295d7]
5: TimeStepper::computeStep()
6: Transient::execute()
7: MooseApp::executeExecutioner()
8: MooseApp::run()
9: main
10: __libc_start_main
11: ../thin_walled-opt(+0x33ef) [0x55ea26ed03ef]
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
```
```
[Variables]
  [disp_x]
    order = FIRST
    family = LAGRANGE
  []
  [disp_y]
    order = THIRD
    family = HERMITE
  []
  [disp_z]
    order = THIRD
    family = HERMITE
  []
  [rot_x]
    order = FIRST
    family = LAGRANGE
  []
  [rot_y]
    order = THIRD
    family = HERMITE
  []
  [rot_z]
    order = THIRD
    family = HERMITE
  []
[]
```

**Reply by GiudGiud** [reply]
that s the solver not converging.
I would recommend going through our 'solver fails' troubleshooting page first
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

**Reply by andrey1887** [reply]
Thank you @GiudGiud . I will try it.

**Reply by andrey1887** [reply]
Hello. @GiudGiud, i runed the Jacobian Debugger for THIRD HERMITE shape functions and got a following:
```
Kernel for variable 'disp_x':
  (0,0) On-diagonal Jacobian is wrong (off by 50.0 %)

Kernel for variable 'disp_y':
  (1,1) On-diagonal Jacobian is questionable (off by 14.85 %)
  (1,5) Off-diagonal Jacobian for variable 'rot_z' is wrong (off by 29.3 %)

Kernel for variable 'disp_z':
  (2,2) On-diagonal Jacobian is questionable (off by 14.85 %)
  (2,4) Off-diagonal Jacobian for variable 'rot_y' is wrong (off by 29.3 %)

Kernel for variable 'rot_x':
  (3,3) On-diagonal Jacobian is wrong (off by 50.0 %)

Kernel for variable 'rot_y':
  (4,2) Off-diagonal Jacobian for variable 'disp_z' is inexact (off by 4.654 %)
  (4,4) On-diagonal Jacobian is wrong (off by 29.3 %)

Kernel for variable 'rot_z':
  (5,1) Off-diagonal Jacobian for variable 'disp_y' is inexact (off by 4.654 %)
  (5,5) On-diagonal Jacobian is wrong (off by 29.3 %)

Kernel for variable 'disp_x':
  (0,0) On-diagonal Jacobian is wrong (off by 50.0 %)

Kernel for variable 'disp_y':
  (1,1) On-diagonal Jacobian is questionable (off by 14.85 %)
  (1,5) Off-diagonal Jacobian for variable 'rot_z' is wrong (off by 29.3 %)

Kernel for variable 'disp_z':
  (2,2) On-diagonal Jacobian is questionable (off by 14.85 %)
  (2,4) Off-diagonal Jacobian for variable 'rot_y' is wrong (off by 29.3 %)

Kernel for variable 'rot_x':
  (3,3) On-diagonal Jacobian is wrong (off by 50.0 %)

Kernel for variable 'rot_y':
  (4,2) Off-diagonal Jacobian for variable 'disp_z' is inexact (off by 4.654 %)
  (4,4) On-diagonal Jacobian is wrong (off by 29.3 %)

Kernel for variable 'rot_z':
  (5,1) Off-diagonal Jacobian for variable 'disp_y' is inexact (off by 4.654 %)
  (5,5) On-diagonal Jacobian is wrong (off by 29.3 %)
```
In case of FIRST LAGRANGE an output is the next:
```
No errors detected. :-)
No errors detected. :-)
```

**Reply by andrey1887** [reply]
i have added output for sizes of `test` and `phi` functions in Jacobian computation process:
```
  std::cout << "_test.size() =" << _test.size();
  std::cout << " _phi.size() =" << _phi.size() << std::endl;
```

```
for (unsigned int i = 0; i < _test.size(); ++i)
  {
    for (unsigned int j = 0; j < _phi.size(); ++j)
    {
      if (_component < 3)
        _local_ke(i, j) = (i == j ? 1 : -1) * _K11[0](_component, _component);
      else
      {
        if (i == j)
          _local_ke(i, j) = _K22[0](_component - 3, _component - 3);
        else
          _local_ke(i, j) = _K22_cross[0](_component - 3, _component - 3);
      }
    }
  }
```
and it was the next:
```
_test.size() =4 _phi.size() =4         # HERMITE 3
_test.size() =2 _phi.size() =2         # LAGRANGE 1
```
I think, that for cubic Hermite approximation, 4-node line element is implied.
And local stiffness matrix becomes 24x24 instead of 12x12 for 2-node element.
Now i see two way to solve this issue (i need cubic approximation for 2-node element):
1. change an algorithm of Jacobian computation
2. change q-rule for element - 2 q-points is enough for 3-order polynomial (but what i can change `_test` function size?)

@GiudGiud, what do you think about it?

**Reply by GiudGiud** [reply]
Changing the Q rule is pretty easy. 

You can look at the Quadrature block in the executioner and specify which one you want

Or you can implement a new one in that same system

**Reply by andrey1887** [reply]
Thank you, @GiudGiud . I try to specify it in the input file:
```
  [./Quadrature]
    type = GAUSS
    order = FIRST         # order = THIRD also tried
  [../]
 ```
 but `_phi` and `_test` -functions sizes haven't changed.

**Reply by GiudGiud** [reply]
yes that s expected. 
I was replying to part of your previous message not the bigger picture of the whole thread. 

You can use any quadrature rule you like, that changed where the equations are evaluated to perform the integration but it doesn't change the shape functions of anything

**Reply by GiudGiud** [reply]
@roystgnr we would like to use a different FE family for the test functions and the variable (different Galerkin method?). How do we do that?

**Reply by andrey1887** [reply]
Ok. Then, maybe the better is to decrease it artificially  in these both cycles:
```
for (unsigned int i = 0; i < _test.size() - 2; ++i)
  {
    for (unsigned int j = 0; j < _phi.size() -2 ; ++j)
    {
```
@GiudGiud , What are you think about it?

**Reply by GiudGiud** [reply]
Maybe the outer one but the inner one would that not leave a bunch of zeros in the Jacobian?
EDIT or the opposite. I need to think about this but I m under-water from another project :S

**Reply by andrey1887** [reply]
Ok, thank you.

**Reply by roystgnr** [reply]
> we would like to use a different FE family for the test functions and the variable 

If this isn't an XY problem: you can always instantiate a `FE<dim,OTHER_FAMILY>`, reinit that on your element, and pull test functions from that.

I'm suspicious, though.  I'd like to see what the weak form looks like.  Petrov-Galerkin methods use different shape functions for the variable and the test function, but in those cases the test functions are more complicated than "comes from a different family", and anyway IMHO it's natural to recast those methods as different variational forms on the same FE space rather than as different bases.  Mixed methods have blocks where a variable of one type is controlled by bilinear form terms with test functions of a completely type, but in those cases the test functions are the variations in a *different* variable in the same system, not a family picked out of the blue.

**Reply by andrey1887** [reply]
Hello, @roystgnr . Thank you. This implementation of line element is based on stiffness matrix, menshioned here [https://github.com/idaholab/moose/discussions/25754#discussioncomment-7537942](url).  So currently i haven't a weak form. But it element based on 6 DOF Euler-Bernoulli beam, for which a third order Hermite polinomials is used (for transverse displacements and rotations while axial disp. and torsion are approximated by the 1 st order shape functions). 
A continuity C1 must be provided. It is becauce i need third order Hermite shape functions. 
In my case it is linear static problem. 
Is there some way to obtain a line element with third-order shape functions but with only 2 nodes and 2 q-points?

**Reply by GiudGiud** [reply]
`2 q-points` : this is separate. Comes from the quadrature.
`only 2 nodes` : @roystgnr

**Reply by roystgnr** [reply]
`HERMITE` is the FE family to use if you need C1 continuity and/or cubics on an EDGE2.  The number of quadrature points is independent of the FE family, though, and by default we use 4 points for a cubic edge, enough to exactly integrate *squares* of cubics, such as you get from time derivative or reaction PDE terms with cubic bases; you'd have to manually reduce quadrature order to go down to 2 ... which I do think would be reasonable if all your shape×test terms are only cubic in one variable and are linear in the other.

I'd have classed elasticity on that beam element as a mixed method; in parts of your matrix you're integrating linear shape functions against cubic test functions, but elsewhere you're integrating cubic shape functions against linear test functions, no?  In that case, no need to make a new `FE`; you just get the cubic test functions from the shape functions evaluated for a transverse displacement or rotation variable.

MOOSE probably has something similar somewhere, but I'm not sure where.  I'd assumed the navier_stokes module would have kernels suitable for the Galerkin form of the Stokes equation on Taylor-Hood elements, but skimming the tests I'm just seeing equal-order (and so presumably pressure-stabilized) input files.

**Reply by andrey1887** [reply]
@roystgnr thank you, but how to decrease number of q-points for 3-rd order Hermite element ?

**Reply by GiudGiud** [reply]
By specifying your own quadrature instead of using the default
https://mooseframework.inl.gov/syntax/Executioner/Quadrature/index.html

**Reply by andrey1887** [reply]
@GiudGiud, i tried the following:
  
  ```
  [Quadrature]
    type = GAUSS
    side_order = FIRST
    element_order = FIRST
    order = FIRST
  []
  ```
  but sizes of shape function don't changed:
  ```
_test.size() =4 _phi.size() =4         # HERMITE 3
_test.size() =2 _phi.size() =2         # LAGRANGE 1
```

**Reply by GiudGiud** [reply]
Just checking something,
These things are double indexed
```
_test[_i][_qp]
```
so you need to index into the shape function (which shape basically) then comes the Qp.
Are you printing that?

**Reply by andrey1887** [reply]
Thank you @GiudGiud .
I did following:
```  
std::cout << "_test.size() =" << _test.size();
std::cout << " _phi.size() =" << _phi.size() << std::endl;

for (_i = 0; _i < _test.size(); ++_i)
    std::cout << _test[_i][_qp] << std::endl;
```
result:
```
_test.size() =2 _phi.size() =2
0
0
_test.size() =4 _phi.size() =4
0
0
0
0
_test.size() =4 _phi.size() =4
0
0
0
0

```
I can't understand how to change the sizes `test` and `phi`

**Reply by GiudGiud** [reply]
The sizes of what you are printing is not the number of quadrature points. It s the number of functions in the basis
try this?
```
std::cout << "Number of qps _test[_i].size() =" << _test[_i].size() << std::endl;
```

**Reply by andrey1887** [reply]
Ok. @GiudGiud , thank you. Anyway a Jacobian size depends on _phi and _test sizes. So i need to change snippets related to them.

**Reply by GiudGiud** [reply]
~~Does it depend on the _test size? Isnt it only the derivatives wrt to the variable DOFs ?~~
nope nevermind there's both

**Reply by andrey1887** [reply]
Hello. @GiudGiud , @roystgnr 
It is about 3-rd Hermite shape functions. So nodal `rot` and  `disp` dofs are used by basis for `disp` variable in 12.10. `rot` then is calculated as it's derivative. Maybe it will clarify the situation.
<img src="https://github.com/idaholab/moose/assets/114737834/d9503f04-3d6f-467f-badb-7ec6ef46d5af" width="600">

**Reply by andrey1887** [reply]
So i have predefined stiffness matrix. Why and how i have to use shape functions despite of they have expressed before in stiffness matrix?

**Reply by GiudGiud** [reply]
@cbolisetti
